Disallow legacy hostNetwork  together with  non-default provider

obnoxxx opened the pull request:
**Description of changes:**

Since the introduction of the "host" network provider, the  legacy
    "hostNetwork" setting is intended to be used only in combination with
    the default network provider (""), but originally, the code did not enforce this.
    
    This change adds the required validation checks to throw errors
    in invalid constellations.





**Issue resolved by this Pull Request:**
Resolves #13692  

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

BlaineEXE commented:
This looks good. Let's also add a validating admission policy to reflect this in `types.go`. That will probably happen in between two of the lines here. And there is an example of how to use the VAPs with `// +kubebuilder:validation:XValidation`.

With that set, it should mean that k8s won't let you set `hostNetwork: true` at the same time as `provider != ""` in the CRD when creating/updating it.

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/network.go:
```diff
@@ -39,6 +39,9 @@ func (n *NetworkSpec) IsHost() bool {
 
 func ValidateNetworkSpec(clusterNamespace string, spec NetworkSpec) error {
+	if spec.HostNetwork && (spec.Provider != NetworkProviderDefault) {
+		return errors.Errorf("the legacy hostNetwork setting is only valid with the default network provider'\"\"' and not with '%q'", spec.Provider)
```
It reads more nicely to use Go's grave-accent string quotes when both double and single quotes are involved.

```suggestion
		return errors.Errorf(1the legacy hostNetwork setting is only valid with the default network provider ("") and not with '%q'`, spec.Provider)
```

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/network_test.go:
```diff
@@ -42,6 +42,23 @@ func TestNetworkCephSpecLegacy(t *testing.T) {
 	assert.Equal(t, expected, net)
 }
 
+func TestValidateNetworkSpec(t *testing.T) {
+	net := NetworkSpec{
+		HostNetwork: true,
+		Provider:    NetworkProviderDefault,
+	}
+	err := ValidateNetworkSpec("", net)
+	assert.Nil(t, err)
+	net = NetworkSpec{
+		HostNetwork: true,
+		Provider:    NetworkProviderHost,
+	}
+	err = ValidateNetworkSpec("", net)
+	//assert.err(t, err)
+	assert.True(t, err != nil)
+
+}
```
Overall, I always want to champion adding unit tests. It's a huge win for reliability, and I am happy to see that you take it seriously! 

I think smaller PRs like this are a really good opportunity to build up some of the baseline knowledge about what Go test semantics tend to look like, as well as the patterns Rook uses internally.

- try to keep subtests visually separated by empty lines for easier at-a-glance reading
- also try to avoid empty lines at the end of methods
- when testing for errors, use testify's `assert.Error(t, err)` or `assert.NoError(t, ett)` to improve at-a-glance reading
- try to caputure the full matrix of value permutations that can arise. I don't see any tests checking what happens when `.HostNetwork == false`, which is an important part of the overall logic and helps check that we don't break users with valid configurations
 
I think I captured all these points here to illustrate.
```suggestion
func TestValidateNetworkSpec(t *testing.T) {
	net := NetworkSpec{
		HostNetwork: true,
		Provider:    NetworkProviderDefault,
	}
	err := ValidateNetworkSpec("", net)
	assert.NoError(t, err)

	net = NetworkSpec{
		HostNetwork: true,
		Provider:    NetworkProviderHost,
	}
	err = ValidateNetworkSpec("", net)
	assert.Error(t, err)
	
	net := NetworkSpec{
		HostNetwork: false,
		Provider:    NetworkProviderDefault,
	}
	err := ValidateNetworkSpec("", net)
	assert.NoError(t, err)
	
	net = NetworkSpec{
		HostNetwork: false,
		Provider:    NetworkProviderHost,
	}
	err = ValidateNetworkSpec("", net)
	assert.NoError(t, err)
}
```

I would also suggest using `t.Run("subtest name", func(/*etc*/){}) to run each test as a subtest. While that isn't really necessary for such a small set of tests, I think it is something that Go devs should know how to do, and learning how to do it here for small, easy tests will help later when your unit tests get more complex.

In a similar vein, I would also encourage you to look into the [Table Driven Tests Pattern](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests). These work great for simple functions like this one, and you'll see them quite a bit throughout Rook's code. Also, FWIW, VSCode can auto-generate a table driven test skeleton automatically as a starting place, which I find hugely helpful. You might find similar tools for your environment.

For `IsHost()`, that generated skeleton looks like this. I haven't modified it at all. It's a little much for what we need here, and I would clean it up a bit for the actual test so the boilerplate isn't so intimidating to read at a glance.

```go
func TestNetworkSpec_IsHost(t *testing.T) {
	type fields struct {
		Provider            NetworkProviderType
		Selectors           map[CephNetworkType]string
		AddressRanges       *AddressRangesSpec
		Connections         *ConnectionsSpec
		HostNetwork         bool
		IPFamily            IPFamilyType
		DualStack           bool
		MultiClusterService MultiClusterServiceSpec
	}
	tests := []struct {
		name   string
		fields fields
		want   bool
	}{
		// TODO: Add test cases.
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			n := &NetworkSpec{
				Provider:            tt.fields.Provider,
				Selectors:           tt.fields.Selectors,
				AddressRanges:       tt.fields.AddressRanges,
				Connections:         tt.fields.Connections,
				HostNetwork:         tt.fields.HostNetwork,
				IPFamily:            tt.fields.IPFamily,
				DualStack:           tt.fields.DualStack,
				MultiClusterService: tt.fields.MultiClusterService,
			}
			if got := n.IsHost(); got != tt.want {
				t.Errorf("NetworkSpec.IsHost() = %v, want %v", got, tt.want)
			}
		})
	}
}
```

obnoxxx commented:
@BlaineEXE  wrote:
> This looks good. Let's also add a validating admission policy to reflect this in `types.go`. That will probably happen in between two of the lines here. And there is an example of how to use the VAPs with `// +kubebuilder:validation:XValidation`.
> 
> With that set, it should mean that k8s won't let you set `hostNetwork: true` at the same time as `provider != ""` in the CRD when creating/updating it.

Thanks for this hint! I found the place in the code. 
It looks a bit as if one should essentially repeat the logic of the Validate function here.

I will soon update the PR accordingly

obnoxxx commented:
@BlaineEXE  wrote:
> This looks good. Let's also add a validating admission policy to reflect this in `types.go`. That will probably happen in between two of the lines here. And there is an example of how to use the VAPs with `// +kubebuilder:validation:XValidation`.
> 
> With that set, it should mean that k8s won't let you set `hostNetwork: true` at the same time as `provider != ""` in the CRD when creating/updating it.

I think I have now added   what you  suggested. And added the updated crds resulting from that.

I guess now it "just" nees some testing ...

obnoxxx commented:
@travisn  wroye:
> Integration tests are failing with the following error, so there must still be some issue with the rule.
> 
> ```
> The CustomResourceDefinition "cephclusters.ceph.rook.io" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[network].properties[provider].x-kubernetes-validations[1].rule: Invalid value: apiextensions.ValidationRule{Rule:"!((self.hostNetwork == true) && (self.provider !=  ''))", Message:"the legacy hostNetwork  setting is only valid if provider is disabled (set to empty  string)", MessageExpression:"", Reason:(*apiextensions.FieldValueErrorReason)(nil), FieldPath:"", OptionalOldSelf:(*bool)(nil)}: compilation failed: ERROR: <input>:1:8: type 'string' does not support field selection | !((self.hostNetwork == true) && (self.provider !=  '')) | .......^ERROR: <input>:1:38: type 'string' does not support field selection | !((self.hostNetwork == true) && (self.provider !=  '')) | 
> ```

Thanks for pointing this out! I have reproduced  the issue locally and think I have a fix now.

Let's see what the CI  yields now ...

obnoxxx commented:
@obnoxxx  wrote:

> Thanks for pointing this out! I have reproduced the issue locally and think I have a fix now.
> 
> Let's see what the CI yields now ...

it seems my fix was incomplete. still failing with the same kind of error at a different place

subhamkrai commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -2320,6 +2320,7 @@ type NetworkSpec struct {
 	// Provider is what provides network connectivity to the cluster e.g. "host" or "multus".
 	// If the Provider is updated from being empty to "host" on a running cluster, then the operator will automatically fail over all the mons to apply the "host" network settings.
 	// +kubebuilder:validation:XValidation:message="network provider must be disabled (reverted to empty string) before a new provider is enabled",rule="self == '' || self == oldSelf"
+	// +kubebuilder:validation:XValidation:message="the legacy hostNetwork  setting can only be set if the network.provider is set to the empty  string",rule="(self.hostNetwork != true) || (self ==  '')"
```
```suggestion
	// +kubebuilder:validation:XValidation:message="the legacy hostNetwork  setting can only be set if the network.provider is set to the empty  string",rule="(self.hostNetwork != true) || (self ==  '')"
```

The rule should be placed before L2319 and probably this`(self.hostNetwork != true) || (self == '')` need to change to `(self.hostNetwork == true) && (self.provider != '')` IMO

subhamkrai replied:
Feel free to go thorugh https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules and https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/ for hints

obnoxxx replied:
@subhamkrai I implemented your suggestion, but ci is still failing in multiple checks in different ways. 

And locally, I can't even get `cluster-test.yaml` created successfully:


here is the bottom of the output of `kubectl describe -n rook-ceph cephclusters my-cluster`: 
```
Events:
  Type     Reason           Age                   From                          Message
  ----     ------           ----                  ----                          -------
  Warning  ReconcileFailed  9m54s (x18 over 20m)  rook-ceph-cluster-controller  failed to reconcile CephCluster "rook-ceph/my-cluster". failed to add finalizer: failed to add finalizer "cephcluster.ceph.rook.io" on "my-cluster": CephCluster.ceph.rook.io "my-cluster" is invalid: spec.network: Invalid value: "object": no such key: hostNetwork evaluating rule: the legacy hostNetwork  setting can only be set if the network.provider is set to the empty string
```
I am surprised to see my newly  added error message here, because the (unmodified)  `cluster-test.yaml` does not even set `hostNetwork` or `provider` at all ...

obnoxxx replied:
@subhamkrai  wrote:
> The rule should be placed before L2319 and probably this`(self.hostNetwork != true) || (self == '')` need to change to `(self.hostNetwork == true) && (self.provider != '')` IMO

Note that `&&` is the wrong logic here: it has to be `||`: We want to declare setting a nonemty provider invalid if `hostNetwork` enabled. Your suggested rule seems to enforce this invalid combination instead.

subhamkrai replied:
I was thinking if both setting is provided through error

obnoxxx replied:
@subhamkrai  wrote:
> I was thinking if both setting is provided through error

It is not clear to me what error you are expecting. The `cluster-test.yaml` was the unmodified version from the rook tree and it contains none of the two settings. So it could only have been an error in code? 🤷🏼

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -2316,6 +2316,7 @@ type SSSDSidecarAdditionalFile struct {
 
 // NetworkSpec for Ceph includes backward compatibility code
 // +kubebuilder:validation:XValidation:message="at least one network selector must be specified when using multus",rule="!has(self.provider) || (self.provider != 'multus' || (self.provider == 'multus' && size(self.selectors) > 0))"
+	// +kubebuilder:validation:XValidation:message="the legacy hostNetwork  setting can only be set if the network.provider is set to the empty string",rule="(self.hostNetwork != true) || (self.provider  ==  '')"
```
I **think** this logic makes sense to me, but @subhamkrai is really the VAP MVP. I would suggest that we try to use `== false` instead of `!= true` since that is a condition that is easier to mentally hold when reading code.

Also, you may need to configure your editor to warn you when inputting double spaces. I keep seeing extra spaces where other contributors don't usually have them.

```suggestion
	// +kubebuilder:validation:XValidation:message="the legacy hostNetwork setting can only be set if the network.provider is set to the empty string",rule="(self.hostNetwork == false) || (self.provider == '')"
```

By my logic, this validates true under the following scenarios, which I think are correct:
- `hostNetwork == false` (and implicitly, when self.provider is any value
- `hostNetwork == true` and `self.provider == ''`

BlaineEXE replied:
This is a line spacing issue. Looking at the CI failure here, it is suggesting a diff to fix the issue.

```
diff pkg/apis/ceph.rook.io/v1/types.go.orig pkg/apis/ceph.rook.io/v1/types.go
--- pkg/apis/ceph.rook.io/v1/types.go.orig
+++ pkg/apis/ceph.rook.io/v1/types.go
@@ -2316,7 +2316,7 @@
 
 // NetworkSpec for Ceph includes backward compatibility code
 // +kubebuilder:validation:XValidation:message="at least one network selector must be specified when using multus",rule="!has(self.provider) || (self.provider != 'multus' || (self.provider == 'multus' && size(self.selectors) > 0))"
-	// +kubebuilder:validation:XValidation:message="the legacy hostNetwork  setting can only be set if the network.provider is set to the empty string",rule="(self.hostNetwork != true) || (self.provider  ==  '')"
+// +kubebuilder:validation:XValidation:message="the legacy hostNetwork  setting can only be set if the network.provider is set to the empty string",rule="(self.hostNetwork != true) || (self.provider  ==  ”)"
 type NetworkSpec struct {
 	// Provider is what provides network connectivity to the cluster e.g. "host" or "multus".
 	// If the Provider is updated from being empty to "host" on a running cluster, then the operator will automatically fail over all the mons to apply the "host" network settings.
```

The line is indented when it should not be.

obnoxxx replied:
Thanks, @BlaineEXE  ! I did the corresponding update, but locally I still get this on the test cluster:

```
Events:
  Type     Reason           Age                From                          Message
  ----     ------           ----               ----                          -------
  Warning  ReconcileFailed  4s (x14 over 45s)  rook-ceph-cluster-controller  failed to reconcile CephCluster "rook-ceph/my-cluster". failed to add finalizer: failed to add finalizer "cephcluster.ceph.rook.io" on "my-cluster": CephCluster.ceph.rook.io "my-cluster" is invalid: spec.network: Invalid value: "object": no such key: hostNetwork evaluating rule: the legacy hostNetwork  setting can only be set if the network.provider is set to the empty string
```

Let's see what the CI thinks ...

subhamkrai replied:
@obnoxxx could you run `make crds` on your branch?

obnoxxx replied:
@subhamkrai  wrote:
> @obnoxxx could you run `make crds` on your branch?

done. and committed the result+pushed

BlaineEXE commented:
@mergifyio rebase

I wonder if this just needs a rebase?

subhamkrai commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -2316,6 +2316,7 @@ type SSSDSidecarAdditionalFile struct {
 
 // NetworkSpec for Ceph includes backward compatibility code
 // +kubebuilder:validation:XValidation:message="at least one network selector must be specified when using multus",rule="!has(self.provider) || (self.provider != 'multus' || (self.provider == 'multus' && size(self.selectors) > 0))"
+// +kubebuilder:validation:XValidation:message="the legacy hostNetwork setting can only be set if the network.provider is set to the empty string",rule="(self.hostNetwork != true) || (self.provider  ==  ”)"
```
```suggestion
// +kubebuilder:validation:XValidation:message="the legacy hostNetwork setting can only be set if the network.provider is set to the empty string",rule="(self.hostNetwork != true) || (self.provider  ==  '')"
```

obnoxxx replied:
@subhamkrai  this suggestion essentially undoes my latest update  which was to fix the build/go vet failure. 
The proposed  change is the inverse of the diff that go vet had produced in the CI and locally ...

obnoxxx replied:
in the meantime, I implemented your suggestion and undid it again since `gofmt` really insists on the strange quote. Additionaly, I added `has() checks to the rule.

Now it  builds just fine locally for me, and I can `kubectl create` the resulting `ceds.yaml` but the created test-cluster from `cluster-test.yaml` still fails to be admitted:

```
Events:
  Type     Reason           Age               From                          Message
  ----     ------           ----              ----                          -------
  Warning  ReconcileFailed  1s (x10 over 4s)  rook-ceph-cluster-controller  failed to reconcile CephCluster "rook-ceph/my-cluster". failed to add finalizer: failed to add finalizer "cephcluster.ceph.rook.io" on "my-cluster": CephCluster.ceph.rook.io "my-cluster" is invalid: spec.network: Invalid value: "object": the legacy hostNetwork setting can only be set if the network.provider is set to the empty string
```

obnoxxx replied:
The CI also shows that gofmt wants the strange quote: https://github.com/rook/rook/actions/runs/7845731244/job/21410846530?pr=13693

BlaineEXE replied:
It seems like we just need to figure out how to resolve this quoting issue. I can surmise that `''` must not be the correct choice for the empty string. 

When I look at the Common Expression Language (the expression syntax used for VAPs) docs, there is a clear definition of what the empty string, as well as how to escape characters. https://github.com/google/cel-spec/blob/master/doc/langdef.md#string-and-bytes-values

I would advise that escape characters be avoided if possible. To this end, Kubebuilder allows using the same golang string syntaxes we are used to, and [reportedly](https://github.com/kubernetes-sigs/controller-tools/issues/315#issuecomment-532407196) we can use the grave accent (or "backtick")(\`) to define the kubebuilder strings.

I'd suggest seeing if the grave accents help, like this:
```
// +kubebuilder:validation:XValidation:message=`msg`,rule=`thing == ""`
```

obnoxxx replied:
@BlaineEXE  I just tried the backticks quoting locally and `make crds` ran fine but produced no diff on the crds. 

Soi I conclude  that the quoting is not the problem with this PR

BlaineEXE commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -2316,6 +2316,7 @@ type SSSDSidecarAdditionalFile struct {
 
 // NetworkSpec for Ceph includes backward compatibility code
 // +kubebuilder:validation:XValidation:message="at least one network selector must be specified when using multus",rule="!has(self.provider) || (self.provider != 'multus' || (self.provider == 'multus' && size(self.selectors) > 0))"
+// +kubebuilder:validation:XValidation:message="the legacy hostNetwork setting can only be set if the network.provider is set to the empty string",rule="(self.hostNetwork != true) || (self.provider == '')"
```
I had to look up CEL documentation to learn that we do need `has()` statements when JSON `omitempty` is used.

The expression here is failing because it does not allow the empty-omitted case to succeed. The expression must be modified to allow `!has(self.hostNetwork)` to be interpreted the same as `self.hostNetwork == false`.

Ditto for `!has(self.provider)` being interpreted the same as `self.provider == ""`.

Please also use `self.hostNetwork == false` instead of `self.hostNetwork != true`. `==` is always easier to read than `!=`.

The parens surrounding the statements between `||` are also unnecessary and don't improve legibility. Let's remove those before final merge.

BlaineEXE replied:
Also, I certainly find that sometimes the required logic needed to explain to VAPs when to pass validation can be hard to grok. I imagine this may be the case for you as well.

I have been finding it easier to mentally write out the failure case and then use [DeMorgan's Laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) to get the inverse expression. I learned this in school, but I'm not sure how common it is in other curriculums.

For example, the failure case is `has(self.hostNetwork) && self.hostNetwork == true && has(self.provider) && self.provider != ""`. The inverse of this is `!(<the whole previous expression>)`, and then DeMorgan's Laws allow simplifying the expression into `||` statements.

obnoxxx replied:
done

obnoxxx replied:
@BlaineEXE  thanks fore the detailed explanation above! I agree with everything you wrote. fwiw, application of Boolean  algebra and de Morgan's law is how I originally arrived at the rule I initially had in the code (involving `||` instead of `&&`.

obnoxxx replied:
After these latest updates, the CI looks a lot better. One failing check looks unrelated.

Locally everything builds and deploys fine. got `cluster-test.yaml` to deploy to status `HEALTH_OK` ! 🎉

BlaineEXE replied:
The inverse of the "failure case" expression I mentioned would be this:

```
!has(self.hostNetwork) || self.hostNetwork == false || !has(self.provider) || self.provider == ""
```

Note that I am using `== false` as the inverse of `== true` and `== ""` as the inverse of `!= ""` to reduce the statement's layers to their simplest boolean algebra forms.

rkachach commented on deploy/charts/rook-ceph/templates/resources.yaml:
```diff
@@ -2282,6 +2282,8 @@ spec:
                   x-kubernetes-validations:
                     - message: at least one network selector must be specified when using multus
                       rule: '!has(self.provider) || (self.provider != ''multus'' || (self.provider == ''multus'' && size(self.selectors) > 0))'
+                    - message: the legacy hostNetwork settingcan only be set if the network.provider is set to the empty string
```
nit: typo "settingcan"

rkachach commented on deploy/examples/crds.yaml:
```diff
@@ -2280,6 +2280,8 @@ spec:
                   x-kubernetes-validations:
                     - message: at least one network selector must be specified when using multus
                       rule: '!has(self.provider) || (self.provider != ''multus'' || (self.provider == ''multus'' && size(self.selectors) > 0))'
+                    - message: the legacy hostNetwork settingcan only be set if the network.provider is set to the empty string
```
nit: typo "settingcan"

obnoxxx replied:
@rkachach  wrote:
> nit: typo "settingcan"

already spotted and fixed.

rkachach commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -2316,6 +2316,7 @@ type SSSDSidecarAdditionalFile struct {
 
 // NetworkSpec for Ceph includes backward compatibility code
 // +kubebuilder:validation:XValidation:message="at least one network selector must be specified when using multus",rule="!has(self.provider) || (self.provider != 'multus' || (self.provider == 'multus' && size(self.selectors) > 0))"
+// +kubebuilder:validation:XValidation:message="the legacy hostNetwork settingcan only be set if the network.provider is set to the empty string",rule="!has(self.hostNetwork) || !has(self.provider)"
```
nit: typo "settingcan"

obnoxxx replied:
@travisn  wrote:
> nit: typo "settingcan"

I had already spotted and fixed this.

parth-gr commented on pkg/apis/ceph.rook.io/v1/network.go:
```diff
@@ -39,6 +39,9 @@ func (n *NetworkSpec) IsHost() bool {
 }
 
 func ValidateNetworkSpec(clusterNamespace string, spec NetworkSpec) error {
+	if spec.HostNetwork && (spec.Provider != NetworkProviderDefault) {
```
do we ever touch this code, as we already have the x-validation for this in the crd?

obnoxxx replied:
@parth-gr  wrote:
> do we ever touch this code, as we already have the x-validation for this in the crd?

good question. This started with the request to adjust this code and then later the request was added to update the x-validation rule. @BlaineEXE , @subhamkrai, wdyt?

BlaineEXE replied:
These validations are still important for 
1. users who don't have a K8s version that supports VAPs
2. any applications that use Rook's API as a library

BlaineEXE replied:
Just because we are checking input values in the beginning of the system, it doesn't mean that checking input values later is not valuable. I have seen others ask this same question in other scenarios, and I want to keep stressing that the idea that "we checked the input once, so we don't need to do it again" is likely to lead to more errors in **any** software system. Input validation is important at all layers of a software system (not just API layers) and is discussed widely in books, blogs, and threads as a good practice. 

Input validation is almost never a computationally expensive operation, and we have almost no reason to do work to avoid it. A large software system can have many places in the code where the values might be modified accidentally or purposefully, and having validation available means that the layer where those values are most important can still verify their validity before it proceeds to configure the system.

The reason this is important for Rook isn't to help the project **now**. It is to help the project throughout its entire maintenance lifespan. If someone makes code changes that affect these values in a seemingly-unrelated portion of the Rook code, Rook's later code paths can still detect an internal configuration error before applying network changes. This likely means that the seemingly-unrelated change will be flagged in CI as causing an error, meaning it won't propagate to end users. Or, if the change does sneak through CI, it at least means that production users won't be affected by the change that otherwise might cause a bug in their Ceph cluster configuration.

parth-gr replied:
Understood, that was a great informatve read.
Thanks

obnoxxx commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -2316,6 +2316,7 @@ type SSSDSidecarAdditionalFile struct {
 
 // NetworkSpec for Ceph includes backward compatibility code
 // +kubebuilder:validation:XValidation:message="at least one network selector must be specified when using multus",rule="!has(self.provider) || (self.provider != 'multus' || (self.provider == 'multus' && size(self.selectors) > 0))"
+// +kubebuilder:validation:XValidation:message="the legacy hostNetwork setting can only be set if the network.provider is set to the empty string",rule="!has(self.hostNetwork) || !has(self.provider)"
```
@parth-gr  wrote:
> rule="!has(self.hostNetwork) || !has(self.provider) -> rule="has(self.hostNetwork) && !has(self.provider)" might be this

Actually no: We want to prevent hostNetwork from being set together with a non-empty provider,  so the rule has to be this:
```
!(has(self.hostNetwork) && has(self.provider))
```
which Boolean algebra simplifies to:
```
!has(self.hostNetwork) || !has(self.provider)
```

obnoxxx replied:
It is actually even more complex since we do need to check the values in addition to using `has()` I'll update shortly.

BlaineEXE reviewed (changes requested):
Looks good, but please squash the commits into one. It's fine to leave the first (`ci: add "network" to the allowed commit message prefixes`) commit separate if you like.

obnoxxx commented:
@BlaineEXE  wrote:
> Looks good, but please squash the commits into one. It's fine to leave the first (`ci: add "network" to the allowed commit message prefixes`) commit separate if you like.

Thanks!

I'll  do the squashing tomorrow

obnoxxx commented:
> @BlaineEXE wrote:
> 
> > Looks good, but please squash the commits into one. It's fine to leave the first (`ci: add "network" to the allowed commit message prefixes`) commit separate if you like.
> 
> Thanks!
> 
> I'll do the squashing tomorrow

I actually already did it now 😬
